### PR TITLE
[SPARK-29145][SQL][FOLLOW-UP] Move tests from `SubquerySuite` to `subquery/in-subquery/in-joins.sql`

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Follow comment of https://github.com/apache/spark/pull/25854#discussion_r342383272

### Why are the changes needed?
NO

### Does this PR introduce any user-facing change?
NO

### How was this patch tested?
ADD TEST CASE

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
@gatorsmile @maropu 
Here if for IN/NOT IN,  Current change can't support `EXISTS/NOT EXISTS` 
They have  another problem to support this. I am working on this.

**Comment by AngersZhuuuu** [comment]
@gatorsmile @maropu 
Currently, use exists in join condition with subquery  have problem with LEFT OUTER JOIN & FULL OUTER JOIN.

**Comment by AngersZhuuuu** [comment]
```

create temporary view s1 as select * from values
    (1), (3), (5), (7), (9)
  as s1(id);

create temporary view s2 as select * from values
    (1), (3), (4), (6), (9)
  as s2(id);

create temporary view s3 as select * from values
    (3), (4), (6), (9)
  as s3(id);

 explain extended SELECT s1.id, s2.id as id2 FROM s1
 LEFT OUTER JOIN s2 ON s1.id = s2.id
 AND EXISTS (SELECT * FROM s3 WHERE s3.id > 6)

```

LogicalPlan 
```
      == Parsed Logical Plan ==
'Project ['s1.id, 's2.id AS id2#291]
+- 'Join LeftOuter, (('s1.id = 's2.id) AND exists#290 [])
   :  +- 'Project [*]
   :     +- 'Filter ('s3.id > 6)
   :        +- 'UnresolvedRelation [s3]
   :- 'UnresolvedRelation [s1]
   +- 'UnresolvedRelation [s2]

== Analyzed Logical Plan ==
id: int, id2: int
Project [id#244, id#250 AS id2#291]
+- Join LeftOuter, ((id#244 = id#250) AND exists#290 [])
   :  +- Project [id#256]
   :     +- Filter (id#256 > 6)
   :        +- SubqueryAlias `s3`
   :           +- Project [value#253 AS id#256]
   :              +- LocalRelation [value#253]
   :- SubqueryAlias `s1`
   :  +- Project [value#241 AS id#244]
   :     +- LocalRelation [value#241]
   +- SubqueryAlias `s2`
      +- Project [value#247 AS id#250]
         +- LocalRelation [value#247]

== Optimized Logical Plan ==
Project [id#244, id#250 AS id2#291]
+- Join LeftOuter, (exists#290 [] AND (id#244 = id#250))
   :  +- Project [value#253 AS id#256]
   :     +- Filter (value#253 > 6)
   :        +- LocalRelation [value#253]
   :- Project [value#241 AS id#244]
   :  +- LocalRelation [value#241]
   +- Project [value#247 AS id#250]
      +- LocalRelation [value#247]

== Physical Plan ==
*(2) Project [id#244, id#250 AS id2#291]
+- *(2) BroadcastHashJoin [id#244], [id#250], LeftOuter, BuildRight, exists#290 []
   :  +- Project [value#253 AS id#256]
   :     +- Filter (value#253 > 6)
   :        +- LocalRelation [value#253]
   :- *(2) Project [value#241 AS id#244]
   :  +- *(2) LocalTableScan [value#241]
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint))), [id=#670]
      +- *(1) Project [value#247 AS id#250]
         +- *(1) LocalTableScan [value#247]
```

Exist can't be change to other form of the join, and `exists` 's child plan was not change to Physical Plan in the whole Physic plan.
```
+- *(2) BroadcastHashJoin [id#244], [id#250], LeftOuter, BuildRight, exists#290 []
   :  +- Project [value#253 AS id#256]
   :     +- Filter (value#253 > 6)
   :        +- LocalRelation [value#253]
```

**Comment by maropu** [comment]
How about moving these tests into `subquery/in-subquery/in-joins.sql` instead of making a new file? Also, I think we need to execute multiple tests runs for these tests with different configurations;
```
--SET spark.sql.autoBroadcastJoinThreshold=10485760
--SET spark.sql.autoBroadcastJoinThreshold=-1,spark.sql.join.preferSortMergeJoin=true
--SET spark.sql.autoBroadcastJoinThreshold=-1,spark.sql.join.preferSortMergeJoin=false
```

**Comment by maropu** [comment]
```
Currently, use exists in join condition with subquery have problem with LEFT OUTER JOIN & FULL OUTER JOIN.
Ok to explain but can't execute.
```
sorry, but I miss your point. You meant that this test could run correctly on `SubquerySuite.scala` but it couldn't run on `SQLQueryTestSuite`?

**Comment by AngersZhuuuu** [comment]
> sorry, but I miss your point. You meant that this test could run correctly on `SubquerySuite.scala` but it couldn't run on `SQLQueryTestSuite`? I think its ok just to move the tests you added in the previous PR into `SQLQueryTestSuite`.

No, current catalyst can't support  well for `EXISTS` when use this as join's `on ` condition and join type is `LeftOuter/FullOuter` like I have show in https://github.com/apache/spark/pull/26406#issuecomment-550159730

In the end of above comment, you can see that some part of Physical Plan is not a Physical Plan , There is a `LocalRelation`, we can't doCodeGen for the whole physical plan.

**Comment by maropu** [comment]
Ur.... I see. You found the bug caused by your previous pr, right? If so, could you file a new jira for that and investigate the root cause of the bug there?

**Comment by AngersZhuuuu** [comment]
> Ur.... I see. You found the bug caused by your previous pr, right? If so, could you file a new jira for that and investigate the root cause of the bug there?

I don't think it is caused by my pr. My pr focus on IN/NOT IN,  behavior between  `IN/NOT IN` and `Exists/NOT EXISTS `  is different.
Run this example in 2.4.0
```
== Parsed Logical Plan ==
'Project ['s1.id, 's2.id AS id2#4]
+- 'Join LeftOuter, (('s1.id = 's2.id) && exists#3 [])
   :  +- 'Project [*]
   :     +- 'Filter ('s3.id > 6)
   :        +- 'UnresolvedRelation `s3`
   :- 'UnresolvedRelation `s1`
   +- 'UnresolvedRelation `s2`

== Analyzed Logical Plan ==
org.apache.spark.sql.AnalysisException: Table or view not found: `s3`; line 3 pos 27;
'Project ['s1.id, 's2.id AS id2#4]
+- 'Join LeftOuter, ((id#0 = id#1) && exists#3 [])
   :  +- 'Project [*]
   :     +- 'Filter ('s3.id > 6)
   :        +- 'UnresolvedRelation `s3`
   :- SubqueryAlias `s1`
   :  +- Project [id#0]
   :     +- SubqueryAlias `s1`
   :        +- LocalRelation [id#0]
   +- SubqueryAlias `s2`
      +- Project [id#1]
         +- SubqueryAlias `s2`
            +- LocalRelation [id#1]

org.apache.spark.sql.AnalysisException: Table or view not found: `s3`; line 3 pos 27;
'Project ['s1.id, 's2.id AS id2#4]
+- 'Join LeftOuter, ((id#0 = id#1) && exists#3 [])
   :  +- 'Project [*]
   :     +- 'Filter ('s3.id > 6)
   :        +- 'UnresolvedRelation `s3`
   :- SubqueryAlias `s1`
   :  +- Project [id#0]
   :     +- SubqueryAlias `s1`
   :        +- LocalRelation [id#0]
   +- SubqueryAlias `s2`
      +- Project [id#1]
         +- SubqueryAlias `s2`
            +- LocalRelation [id#1]

== Optimized Logical Plan ==
org.apache.spark.sql.AnalysisException: Table or view not found: `s3`; line 3 pos 27;
'Project ['s1.id, 's2.id AS id2#4]
+- 'Join LeftOuter, ((id#0 = id#1) && exists#3 [])
   :  +- 'Project [*]
   :     +- 'Filter ('s3.id > 6)
   :        +- 'UnresolvedRelation `s3`
   :- SubqueryAlias `s1`
   :  +- Project [id#0]
   :     +- SubqueryAlias `s1`
   :        +- LocalRelation [id#0]
   +- SubqueryAlias `s2`
      +- Project [id#1]
         +- SubqueryAlias `s2`
            +- LocalRelation [id#1]

== Physical Plan ==
org.apache.spark.sql.AnalysisException: Table or view not found: `s3`; line 3 pos 27;
'Project ['s1.id, 's2.id AS id2#4]
+- 'Join LeftOuter, ((id#0 = id#1) && exists#3 [])
   :  +- 'Project [*]
   :     +- 'Filter ('s3.id > 6)
   :        +- 'UnresolvedRelation `s3`
   :- SubqueryAlias `s1`
   :  +- Project [id#0]
   :     +- SubqueryAlias `s1`
   :        +- LocalRelation [id#0]
   +- SubqueryAlias `s2`
      +- Project [id#1]
         +- SubqueryAlias `s2`
            +- LocalRelation [id#1]
Time taken: 1.455 seconds, Fetched 1 row(s)
```

**Comment by AngersZhuuuu** [comment]
> Ur.... I see. You found the bug caused by your previous pr, right? If so, could you file a new jira for that and investigate the root cause of the bug there?

My pr make this situation pass Optimizer level. == , Anyway, I will raise a issue and explain  the change and root cause. 
Also try to make it work.

**Comment by maropu** [comment]
If so, could you separate the two work: the porting task you requested and the bug fix? I'm pretty confused now.

**Comment by AngersZhuuuu** [comment]
> If so, could you separate the two work: the porting task you requested and the bug fix? I'm pretty confused now.

Yeah, I want to make it clear to you since in comment https://github.com/apache/spark/pull/25854#discussion_r342383272  @gatorsmile  mentioned two problem together.  I will raise a issue for `EXISTS/NOT EXISTS` problem.

**Comment by AngersZhuuuu** [comment]
> How about moving these tests into `subquery/in-subquery/in-joins.sql` instead of making a new file? Also, I think we need to execute multiple runs for these tests with different configurations;
> 
> ```
> --SET spark.sql.autoBroadcastJoinThreshold=10485760
> --SET spark.sql.autoBroadcastJoinThreshold=-1,spark.sql.join.preferSortMergeJoin=true
> --SET spark.sql.autoBroadcastJoinThreshold=-1,spark.sql.join.preferSortMergeJoin=false
> ```

move to in-join.sql,  it will be run in three configuration auto?

**Comment by maropu** [comment]
Yea, you need to nothing for that. btw, still `WIP` in this pr?

**Comment by AngersZhuuuu** [comment]
> Yea, you need to nothing for that. btw, still `WIP` in this pr?

Remove `WIP`

**Comment by maropu** [comment]
Can you remove the test of `SubquerySuite` that you added in the previous pr? https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala#L207-L353

**Comment by AngersZhuuuu** [comment]
> Can you remove the test of `SubquerySuite` that you added in the previous pr? https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala#L207-L353

Done

File: sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-joins.sql [hunk 1]
```diff
@@ -51,6 +51,18 @@ create temporary view t3 as select * from values
...
+  as s1(id);
+
+create temporary view s2 as select * from values
+    (1), (3), (4), (6), (9)
+  as s2(id);
+
+create temporary view s3 as select * from values
+    (3), (4), (6), (9)
+  as s3(id);
+
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-joins.sql]: Plz drop view in the end?

**Comment by maropu** [comment]
ok to test

File: sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-joins.sql (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/inputs/subquery/in-subquery/in-joins.sql]: > Plz drop view in the end?

About t1, t2, t3, seems in other test file, don't drop table.

**Comment by AngersZhuuuu** [comment]
@maropu Passed test, any more work to do?

**Review by maropu — APPROVED** [review]
cc: @dongjoon-hyun @HyukjinKwon

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @maropu .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.

**Comment by maropu** [comment]
Thanks, @dongjoon-hyun !

**Pull request closed without merging**